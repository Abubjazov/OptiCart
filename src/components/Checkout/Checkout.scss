@import '../../variables';

.checkout {
	background-color: $white;
	display: flex;
	flex-direction: column;
	align-items: flex-end;

	@media (max-width: 930px) {
		margin-top: 10px;
		flex-direction: row;
		justify-content: space-between;
	}

	@media (max-width: 400px) {
		margin-top: 2px;
	}

	svg {
		margin: 0 75px;

		@media (max-width: 440px) {
			margin: 0 55px;
		}
	}

	h2 {
		margin-right: 25px;
		font-size: 24px;
		line-height: 45px;
		font-weight: 300;

		@media (max-width: 400px) {
			font-size: 20px;
			margin-right: 0;
		}
	}

	button {
		background-color: $main-color;
		border: 1px solid $gray-light;
		border-style: none none solid none;
		color: $white;
		font-size: 20px;
		font-weight: 300;
		padding: 10px 0;
		width: 200px;
		transition: all 0.2s linear;
		cursor: pointer;

		@media (max-width: 400px) {
			width: 170px;
			padding: 8px 0;
		}

		&:hover {
			background: $deep-main;
		}
	}
}
