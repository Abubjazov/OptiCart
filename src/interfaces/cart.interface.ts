import { Product } from './product.interface'

export enum CartActionTypes {
	FETCH_CART = 'FETCH_CART',
	FETCH_CART_SUCCESS = 'FETCH_CART_SUCCESS',
	FETCH_CART_ERROR = 'FETCH_CART_ERROR',
	ADD_TO_CART = 'ADD_TO_CART',
	ADD_TO_CART_SUCCESS = 'ADD_TO_CART_SUCCESS',
	ADD_TO_CART_ERROR = 'ADD_TO_CART_ERROR',
	REMOVE_FROM_CART = 'REMOVE_FROM_CART',
	REMOVE_FROM_CART_SUCCESS = 'REMOVE_FROM_CART_SUCCESS',
	REMOVE_FROM_CART_ERROR = 'REMOVE_FROM_CART_ERROR',
	UPDATE_CART_QUANTITY = 'UPDATE_CART_QUANTITY',
	UPDATE_CART_QUANTITY_SUCCESS = 'UPDATE_CART_QUANTITY_SUCCESS',
	UPDATE_CART_QUANTITY_ERROR = 'UPDATE_CART_QUANTITY_ERROR',
}

export interface CartListItem extends Product {
	product_id: number
	quantity: number
}

export interface CartState {
	cart: CartListItem[]
	currentItemId: null | number
	status: 'waiting' | 'error' | 'loading'
	error: null | string
}

interface FetchCartAction {
	type: CartActionTypes.FETCH_CART
}

interface FetchCartSuccessAction {
	type: CartActionTypes.FETCH_CART_SUCCESS
	payload: CartListItem[]
}

interface FetchCartErrorAction {
	type: CartActionTypes.FETCH_CART_ERROR
	payload: string
}

interface addToCartAction {
	type: CartActionTypes.ADD_TO_CART
	payload: number
}

interface addToCartSuccessAction {
	type: CartActionTypes.ADD_TO_CART_SUCCESS
	payload: CartListItem[]
}
interface addToCartErrorAction {
	type: CartActionTypes.ADD_TO_CART_ERROR
	payload: string
}
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }

export type CartAction =
	| FetchCartAction
	| FetchCartSuccessAction
	| FetchCartErrorAction
	| addToCartAction
	| addToCartSuccessAction
	| addToCartErrorAction
