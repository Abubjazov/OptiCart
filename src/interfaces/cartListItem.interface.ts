import { Product } from './product.interface'

export enum CartActionTypes {
	FETCH_CART = 'FETCH_CART',
	FETCH_CART_SUCCESS = 'FETCH_CART_SUCCESS',
	FETCH_CART_ERROR = 'FETCH_CART_ERROR',
}

export interface CartListItem extends Product {
	product_id: number
	quantity: number
}

export interface CartState {
	cart: CartListItem[]
	status: 'waiting' | 'error' | 'loading'
	error: null | string
}

interface FetchCartAction {
	type: CartActionTypes.FETCH_CART
}

interface FetchCartSuccessAction {
	type: CartActionTypes.FETCH_CART_SUCCESS
	payload: CartListItem[]
}

interface FetchCartErrorAction {
	type: CartActionTypes.FETCH_CART_ERROR
	payload: string
}

export type CartAction =
	| FetchCartAction
	| FetchCartSuccessAction
	| FetchCartErrorAction
