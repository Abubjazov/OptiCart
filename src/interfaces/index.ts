export * from './cartListItem.interface'
export * from './product.interface'
export * from './action.interface'
