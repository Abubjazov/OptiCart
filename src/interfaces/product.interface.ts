export enum ProductActionTypes {
	FETCH_PRODUCTS = 'FETCH_PRODUCTS',
	FETCH_PRODUCTS_SUCCESS = 'FFETCH_PRODUCTS_SUCCESS',
	FETCH_PRODUCTS_ERROR = 'FETCH_PRODUCTS_ERROR',
}

export interface Product {
	id: number
	name: string
	picture: string
	picture_medium: string
	picture_small: string
	description: string
	price: number
}

export interface ProductState {
	products: Product[]
	currentProductId: null | number
	status: 'waiting' | 'error' | 'loading'
	error: null | string
}

interface FetchProductsAction {
	type: ProductActionTypes.FETCH_PRODUCTS
}

interface FetchProductsSuccessAction {
	type: ProductActionTypes.FETCH_PRODUCTS_SUCCESS
	payload: Product[]
}

interface FetchProductsErrorAction {
	type: ProductActionTypes.FETCH_PRODUCTS_ERROR
	payload: string
}

export type ProductAction =
	| FetchProductsAction
	| FetchProductsSuccessAction
	| FetchProductsErrorAction
