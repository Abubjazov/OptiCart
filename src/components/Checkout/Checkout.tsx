import { useActions } from '../../hooks/useActions'
import { useTypedSelector } from '../../hooks/useTypedSelector'
import { getTotal } from '../../utils/utils'

import { MediumSpinnerBlack } from '../Spinners/MediumSpinnerBlack'

import './Checkout.scss'

export const Checkout = (): JSX.Element => {
	const { cart, status } = useTypedSelector(state => state.cart)
	const { checkoutCart } = useActions()

	return (
		<section className='checkout'>
			{status === 'loading' ? (
				<MediumSpinnerBlack />
			) : (
				<h2>Total: {getTotal(cart)} $</h2>
			)}

			<button onClick={checkoutCart}>Checkout</button>
		</section>
	)
}
