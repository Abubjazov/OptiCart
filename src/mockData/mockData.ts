import { CartListItem } from '../interfaces'

export const cartItems: CartListItem[] = [
	{
		id: 2,
		product_id: 15,
		name: 'Rangoto Vi',
		picture: 'https://i.ibb.co/T2XSnw0/z9.webp',
		description: 'Nemo dolore officia est natus.',
		price: 11.04,
		quantity: 3,
	},
	{
		id: 3,
		product_id: 16,
		name: 'Armo Sento 2',
		picture: 'https://i.ibb.co/TPT93hh/z10.jpg',
		description: 'Nobis in voluptatum qui voluptate.',
		price: 17.22,
		quantity: 1,
	},
	{
		id: 4,
		product_id: 9,
		name: 'Vertigo 2',
		picture: 'https://i.ibb.co/vvCZ6gp/nike.webp',
		description: 'Architecto omnis sed nostrum aliquid.',
		price: 26.11,
		quantity: 2,
	},
]
