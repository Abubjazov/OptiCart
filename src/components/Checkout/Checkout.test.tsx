import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { CartActionTypes } from '../../interfaces'

import { mockCartItems } from '../../mockData/mockData'

import { Checkout } from './Checkout'

const mockStore = configureStore([])

describe('Component: Checkout', () => {
	test('should render Checkout with cart.status = "loading"', () => {
		const store = mockStore({
			cart: {
				cart: [],
				status: 'loading',
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<Checkout />
			</Provider>
		)
		expect(asFragment()).toMatchSnapshot()
	})

	test('should render Checkout with default store state', () => {
		const store = mockStore({
			cart: {
				cart: [],
				status: 'waiting',
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<Checkout />
			</Provider>
		)
		expect(asFragment()).toMatchSnapshot()
	})

	test('should render Checkout with mockData store state', () => {
		const store = mockStore({
			cart: {
				cart: mockCartItems,
				status: 'waiting',
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<Checkout />
			</Provider>
		)
		expect(asFragment()).toMatchSnapshot()
	})

	test('should dispatch an action on "Checkout" button click', async () => {
		const middlewares = [thunk]
		const mockStore = configureStore(middlewares)

		let store = mockStore({
			cart: {
				cart: mockCartItems,
				status: 'waiting',
			},
		})

		const expectedActions = [{ type: CartActionTypes.CHECKOUT_CART }]

		render(
			<Provider store={store}>
				<Checkout />
			</Provider>
		)

		userEvent.click(screen.getByText('Checkout'))
		expect(store.getActions()).toEqual(expectedActions)
	})
})
