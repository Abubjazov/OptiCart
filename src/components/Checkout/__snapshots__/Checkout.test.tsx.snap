// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Checkout should render Checkout 1`] = `
<DocumentFragment>
  <section
    class="checkout"
  >
    <h2>
      Total: 0.00 $
    </h2>
    <button>
      Checkout
    </button>
  </section>
</DocumentFragment>
`;
